Reformation of perf suite latency tests

wrrobin opened the pull request:
1. Normalizing perf metrics between bandwidth and latency tests
2. Guarded THREAD* constants with ifdef flags
3. Multiple PE latency tests
4. Code re-organization and modularizing
5. Multi-threaded latency tests (TBD)

jdinan commented on test/performance/shmem_perf_suite/bw_common.h:
```diff
@@ -747,43 +339,48 @@ static inline void uni_dir_bw_test_and_output(perf_metrics_t metric_info) {
-    }
 
 #ifndef VERSION_1_0
+#if defined(ENABLE_THREADS)
```
Where is `ENABLE_THREADS` coming from?

jdinan replied:
Did you mean `ENABLE_OPENMP`?

jdinan replied:
Never mind, this is added to AM_CPPFLAGS in `Makefile.am`.

jdinan commented on test/performance/shmem_perf_suite/bw_common.h:
```diff
@@ -747,43 +339,48 @@ static inline void uni_dir_bw_test_and_output(perf_metrics_t metric_info) {
-        bar_psync[i] = SHMEM_SYNC_VALUE;
+    if (error_checking_init_target_usage(metric_info) == -1)
+        return -1;
+#if defined(ENABLE_THREADS)
```
`ENABLE_THREADS`?

jdinan commented on test/performance/shmem_perf_suite/common.h:
```diff
@@ -47,6 +47,148 @@
+typedef enum {
+    B,
+    KB,
+    MB
```
Should these have specific values, e.g. B=1, KB=1024, etc?

wrrobin replied:
This ``enumerator`` type is used to just detect how the output bandwidth is to be reported. Based on the enum values, we multiply with the appropriate value. We can go either way for this.

wrrobin replied:
@jdinan, Keeping this as it is for now. Will refine this in a future PR.

jdinan commented on test/performance/shmem_perf_suite/common.h:
```diff
@@ -47,6 +47,148 @@
+    metric_info->src = NULL;
+    metric_info->dest = NULL;
+
+#if defined(ENABLE_THREADS)
```
`ENABLE_THREADS`?

jdinan commented on test/performance/shmem_perf_suite/common.h:
```diff
@@ -123,41 +264,40 @@ static void aligned_buffer_free(char * ptr_aligned)
+int is_divisible_by_4(int num) {
     assert(num >= 0);
     assert(sizeof(int) == 4);
     return (!(num & 0x00000003));
```
This is not portable, because C doesn't require a particular representation for signed integers.  Unless this is in an innermost loop (looks like it's only used in diagnostic code), you can safely write `num % 4 == 0` and rely on the compiler to do the right thing.  Also, the assert non-negative should probably be `if (num < 0) return 0;` since assert should not be used for a persistent error check.

wrrobin replied:
Ok. Will update this.

jdinan replied:
Actually, `shmem_global_exit` may be preferred instead of `return 0`.

jdinan commented on test/performance/shmem_perf_suite/common.h:
```diff
@@ -167,3 +307,466 @@ static inline void validate_recv(char * buf, int len, int partner_pe)
+            metric_info->szinitiator = strtoul(optarg, (char **)NULL, 0);
+            break;
+        case 'C':
+#if defined(ENABLE_THREADS)
```
ENABLE_THREADS?

jdinan commented on test/performance/shmem_perf_suite/shmem_latency_get_ctx_perf.c:
```diff
@@ -0,0 +1,53 @@
+**
+**  -in micro seconds
+**
+**NOTE: this test assumes correctness of reduction algorithm
```
@kseager Can we remove this comment?

kseager replied:
yes

wrrobin replied:
ok, will do.

kseager commented on test/performance/shmem_perf_suite/latency_common.h:
```diff
@@ -31,57 +31,31 @@
-   printf("\nLength                  Latency                       \n");
-   printf("in bytes            in micro seconds              \n");
+static inline 
+void print_latency_header(perf_metrics_t metric_info) {
```
no need to include metric_info for this call

wrrobin replied:
Thanks @kseager. Will remove it for now. 
I kept it initially to print with min, max, and avg options, if we plan to add it. But, this PR is out of scope for that.

wrrobin commented on test/performance/shmem_perf_suite/latency_ctx.h:
```diff
@@ -0,0 +1,203 @@
+    unsigned long int i;
+    int dest = partner_node(*metric_info);
+    char *src = aligned_buffer_alloc(metric_info->nthreads * len);
+    char *dst = aligned_buffer_alloc(metric_info->nthreads * len);
```
Sure, will do.

wrrobin commented:
@jdinan @kseager This PR is ready for a re-review. Some additional changes apart from your previous suggestions are:
1. Fixed validation issue for latency tests
2. Added opstyle in latency tests to figure out the appropriate validation node in latency tests
3. Removed separate buffers in MT tests and utilized the pre-allocated ones in perf-metrics.

wrrobin commented:
Thanks @kseager @jdinan. Going with the merge then.
